Controller/Certificates/Download.php

```diff
@@ -0,0 +1,158 @@
+ *
+ * @copyright  Copyright (c) 2018 Avalara, Inc.
+ * @license    http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+
+namespace ClassyLlama\AvaTax\Controller\Certificates;
+
+use ClassyLlama\AvaTax\Framework\Interaction\Rest\Customer;
+use ClassyLlama\AvaTax\Helper\UrlSigner;
+use Magento\Customer\Api\CustomerRepositoryInterface;
+use Magento\Framework\App\RequestInterface;
+use Magento\Framework\DataObjectFactory;
+use Magento\Framework\Exception\NotFoundException;
+
+class Download extends \Magento\Framework\App\Action\Action
+{
+    /**
+     * @var \Magento\Framework\Controller\Result\RawFactory
+     */
+    protected $resultRawFactory;
+
+    /**
+     * @var \Magento\Framework\App\Response\Http\FileFactory
+     */
+    protected $fileFactory;
+
+    /**
+     * @var Customer
+     */
+    protected $customerRest;
+
+    /**
+     * @var UrlSigner
+     */
+    protected $urlSigner;
+
+    /**
+     * @var CustomerRepositoryInterface
+     */
+    protected $customerRepository;
+
+    /**
+     * @var DataObjectFactory
+     */
+    protected $dataObjectFactory;
+
+    /**
+     * @var \Magento\Customer\Model\Session
+     */
+    protected $session;
+
+    /**
+     * Download constructor.
+     *
+     * @param \Magento\Framework\Controller\Result\RawFactory  $resultRawFactory
+     * @param \Magento\Framework\App\Response\Http\FileFactory $fileFactory
+     * @param Customer                                         $customerRest
+     * @param UrlSigner                                        $urlSigner
+     * @param CustomerRepositoryInterface                      $customerRepository
+     * @param DataObjectFactory                                $dataObjectFactory
+     * @param \Magento\Backend\App\Action\Context              $context
+     * @param \Magento\Customer\Model\Session                  $session
+     */
+    public function __construct(
+        \Magento\Framework\Controller\Result\RawFactory $resultRawFactory,
+        \Magento\Framework\App\Response\Http\FileFactory $fileFactory,
+        Customer $customerRest,
+        UrlSigner $urlSigner,
+        CustomerRepositoryInterface $customerRepository,
+        DataObjectFactory $dataObjectFactory,
+        \Magento\Backend\App\Action\Context $context,
+        \Magento\Customer\Model\Session $session
+    )
+    {
+        parent::__construct($context);
+
+        $this->resultRawFactory = $resultRawFactory;
+        $this->fileFactory = $fileFactory;
+        $this->customerRest = $customerRest;
+        $this->urlSigner = $urlSigner;
+        $this->customerRepository = $customerRepository;
+        $this->dataObjectFactory = $dataObjectFactory;
+        $this->session = $session;
+    }
+
+    /**
+     * @param RequestInterface $request
+     *
+     * @return \Magento\Framework\App\ResponseInterface
+     * @throws \Magento\Framework\Exception\NotFoundException
+     */
+    public function dispatch(RequestInterface $request)
```

Comment:
@LordZardeck - This method begins to illustrate why I'm uncomfortable with this controller being directly used by the admin controller, because it's not intuitive when looking at the code that this authentication check (which wouldn't be applicable in the admin) isn't ever performed when this controller is being used in the other context (because it's not being used as a controller there, it's just having "execute" called directly).  Blocks and controllers should be used as blocks and controllers, not as generic dependencies.